apps/api/src/report/report.service.ts

```diff
@@ -347,81 +378,130 @@ export class ReportService {
+   *    First Nations Health Authority
+   *    Fraser Health Authority
+   *    Interior Health Authority
+   *    Northern Health Authority
+   *    Providence Health Care Society
+   *    Provincial Health Services Authority
+   *    Vancouver Coastal Health Authority
+   *    Vancouver Island Health Authority
+   * Immigration
+   * Overall
+   * Average time to hire
+   *
+   * @param end date of report, YYYY-MM-DD
+   * @param verbose if true, include the list of values used to calculate mean, mode, and median.
    */
-  async getAverageTimeWithEachStakeholderGroup(statuses: Record<string, string>, t: string) {
-    const { to } = this.captureFromTo('', t);
-    this.logger.log(`getAverageTimeWithEachStakeholderGroup: apply filter till ${to} date)`);
-
-    // to get durations of NNAS, BCCNM & NCAS, Immigration, Overall, and Average time to hire
-    const durations = await getRepository(MilestoneDurationEntity)
-      .createQueryBuilder()
-      .where(`hired_at <= :to AND immigrated_at  <= :to`, { to })
-      .getMany();
-
-    // excludes applicants with a negative duration caused by nonlinear milestones
-    const linearDurations = durations.filter(d => !Object.values(d).some(v => v < 0));
-
-    // to get duration of recruitment by Health Authorities
-    const durationByHAs = await getManager()
-      .createQueryBuilder(IENHaPcn, 'ha')
-      .select('ha.title', 'HA')
-      .addSelect(`CAST(COALESCE(ROUND(AVG(md.recruitment), 2), 0) AS FLOAT)`, 'Mean')
-      .addSelect(`COALESCE(MODE() WITHIN GROUP (ORDER BY md.recruitment), 0)`, 'Mode')
-      .addSelect(
-        `COALESCE(PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY md.recruitment), 0)`,
-        'Median',
-      )
-      .leftJoin(
-        qb => {
-          return qb
-            .from(MilestoneDurationEntity, 'd')
-            .where(`d.hired_at <= :to AND d.immigrated_at <= :to`, { to })
-            .andWhere(`LEAST(d.pre_screen, interview, reference_check, hired) > 0`);
-        },
-        'md',
-        'ha.title = md.ha',
+  async getReport9(end: string, verbose = false): Promise<object[]> {
+    const { to } = this.captureFromTo('', end);
+    this.logger.log(
+      `Report 9 - Average time with each health authority: apply filter till ${to} date)`,
+      'REPORT',
+    );
+    const table = await this.reportUtilService.getMilestoneTable(to);
+
+    // set health authorities who hired applicants
+    const idHaMap = await this.reportUtilService.getHiredApplicantHAs();
+    for (const [id, ha] of Object.entries(idHaMap)) {
+      if (table[id]) table[id].ha = ha;
+    }
+
+    // calculate durations and summaries
+    const durations = Object.values(table)
+      .map(this.reportUtilService.getDurationTableEntry)
+      .map(this.reportUtilService.getDurationSummary);
+
+    // generate final report form
+    const data = _.flatten(
+      DURATION_STAGES.map(stage => {
+        const applicants = durations.filter(d => d[stage[0].stage as keyof DurationSummary]);
```

Comment:
I am finding `stage[0].stage` hard to understand without investigating. Since its used in a few places, maybe assign it to a variable with a descriptive name.

Reply:
moved the position of stageName definition.